mariadb: 10.3.22 -> 10.4.12

Izorkin opened the pull request:
###### Motivation for this change
Upgrade MariaDB server to 10.4.12

cc @thoughtpolice @flokli @aanderse @dasJ
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Izorkin committed:
mariadb-galera: 25.3.27 -> 26.4.3

Izorkin committed:
mariadb: 10.3.22 -> 10.4.12

Izorkin committed:
nixos/mysql: disable load pluginx auth_socket in mariadb

Izorkin committed:
nixos/mysql: add release notes

aanderse reviewed (commented):
Thank you for doing this @Izorkin!

flokli commented on pkgs/servers/sql/mariadb/cmake-disable-auth-pam-testing.patch:
```diff
@@ -0,0 +1,12 @@
+                    COMPONENT Server)
+   ENDIF()
+   IF(TARGET auth_pam OR TARGET auth_pam_v1)
+-    ADD_SUBDIRECTORY(testing)
```
Please don't just simply disable tests, or worse, add patches disabling tests without any comment on why this is necessary.

Izorkin replied:
Added comment.

flokli commented on nixos/modules/services/databases/mysql.nix:
```diff
@@ -87,7 +87,6 @@ in
             datadir = /var/lib/mysql
             bind-address = 127.0.0.1
             port = 3336
-            plugin-load-add = auth_socket.so
```
Is `auth_socket.so` now included?

aanderse replied:
Yes.

Izorkin commented on pkgs/servers/sql/mariadb/cmake-disable-auth-pam-testing.patch:
```diff
@@ -0,0 +1,12 @@
+                    COMPONENT Server)
+   ENDIF()
+   IF(TARGET auth_pam OR TARGET auth_pam_v1)
+-    ADD_SUBDIRECTORY(testing)
```
Singly. This patch remove directories `./result/OFF/suite/plugins/pam` and files `pam_mariadb_mtr.so`, `mariadb_mtr`. https://jira.mariadb.org/browse/MDEV-21654

Izorkin commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -61,6 +61,16 @@
    </listitem>
+   <listitem>
+    <para>
+      MariaDB updated to 10.4 and plugin MariaDB Galera to 26.4.
```
Fixed, thanks.

flokli reviewed (commented):
LGTM, but haven't tested it - especially the migration part, so would love to get some more feedback from people using this.

Izorkin commented:
@GrahamcOfBorg test mariadb-galera-mariabackup

Izorkin commented:
@GrahamcOfBorg test mysql-replication

Izorkin commented:
Found errors with authentication.
https://mariadb.com/kb/en/authentication-from-mariadb-104

Izorkin committed:
nixos/tests/mariadb-galera-mariabackup: fix test script

MariaDB 10.4 introduces a number of changes to the authentication
process, intended to make things easier and more intuitive.

Izorkin committed:
nixos/tests/mariadb-galera-rsync: fix test script

MariaDB 10.4 introduces a number of changes to the authentication
process, intended to make things easier and more intuitive.

Izorkin committed:
nixos/tests/mysql-replication: fix test script

MariaDB 10.4 introduces a number of changes to the authentication
process, intended to make things easier and more intuitive.

Izorkin committed:
nixos/mysql-replication: ignore system databases in binary log

Izorkin commented:
@GrahamcOfBorg test mariadb-galera-mariabackup
@GrahamcOfBorg test mariadb-galera-rsync
@GrahamcOfBorg test mysql
@GrahamcOfBorg test mysql-autobackup
@GrahamcOfBorg test mysql-backup
@GrahamcOfBorg test mysql-replication

aanderse commented:
Test is failing on `arm` because `TokuDB` is unsupported plugin it looks like.

@flokli I'll upgrade a box or two this weekend to test this out, but we should be looking to merge early enough that our `unstable` users can catch any issues before `20.09`.

Thank you very much for your hard work @Izorkin!

flokli commented:
it seems `ha_tokudb.so` isn't part of the mysql output path:

> [ERROR] mysqld: Can't open shared library '/nix/store/al7j8gzl8qjhh6fdiwjdh7bqyrrm911r-mariadb-server-10.4.12/lib/mysql/plugin/ha_tokudb.so' (errno: 22, cannot open shared object file: No such file or directory)

However, this is already the case on master (aarch64 and i686).

We might want to figure out why it's not available on these arches, and possibly update the test to only test `tokudb` where it's available, but this can be done in a follow-up PR.

aanderse commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -61,6 +61,18 @@
+      Before you upgrade, it would be best to take a backup of your database.
+      For MariaDB Galera Cluster, see <link xlink:href="https://mariadb.com/kb/en/upgrading-from-mariadb-103-to-mariadb-104-with-galera-cluster/">Upgrading
+      from MariaDB 10.3 to MariaDB 10.4 with Galera Cluster</link> instead.
+      Before doing the upgrade read <link xlink:href="https://mariadb.com/kb/en/upgrading-from-mariadb-103-to-mariadb-104/#incompatible-changes-between-103-and-104">Incompatible
```
It might be worth explicitly stating (in bold) the user will need to run `mysql_upgrade` after upgrading, even though this is included in upstream documentation. Since NixOS is not entirely consistent about running migrations some users may assume this was already done for them.

Izorkin replied:
Updated release note and rebased PR.

aanderse reviewed (approved):
I upgraded a machine to this package, ran `mysql_upgrade`, and then tested some connections and programs out... everything worked fine :+1:

Thanks again @Izorkin :tada:

aanderse commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -61,6 +61,19 @@
+      Changes Between 10.3 and 10.4</link>.
+      MariaDB 10.4 introduces a number of changes to the authentication process, intended to make things easier and more
+      intuitive. See <link xlink:href="https://mariadb.com/kb/en/authentication-from-mariadb-104/">Authentication from MariaDB 10.4</link>.
+      After upgrade need run <literal>mysql_upgrade</literal>.
```
```suggestion
      After the upgrade you will need to run <literal>mysql_upgrade</literal>.
```

Izorkin replied:
Fixed.

aanderse reviewed (approved):
@flokli I suggest merge once the bot confirms everything. Any issue with that?

flokli commented on pkgs/top-level/all-packages.nix:
```diff
@@ -15847,8 +15847,8 @@ in
-  mariadb-galera = mariadb-galera_25;
-  mariadb-galera_25 = callPackage ../servers/sql/mariadb/galera/25.nix {
+  mariadb-galera = mariadb-galera_26;
+  mariadb-galera_26 = callPackage ../servers/sql/mariadb/galera/26.nix {
```
Why do we provide `mariadb-galera_*` here?

What's the usecase? If we want to provide users with a way to stick to the old version, then it shouldn't be removed.

If we don't provide this option, there's not much of a point in a "versioned" attribute.

I'd propose to just leave the `mariadb-galera` attribute, and add an alias to `aliases.nix`.

Izorkin replied:
Rebased. Only one version left.

flokli commented on nixos/tests/mysql/mariadb-galera-mariabackup.nix:
```diff
@@ -184,40 +184,40 @@ in {
     galera_03.wait_for_unit("mysql")
     galera_03.wait_for_open_port(3306)
     galera_02.succeed(
-        "sudo -u testuser mysql -u root -e 'use testdb; select test_id from db1;' -N | grep 37"
```
I'd need some context here. Were these tests broken before this PR? Are these changes needed because of the version bump?

In both cases, please add the explanation to the commit message.

Izorkin replied:
MariaDB 10.4 introduces a number of changes to the authentication process, intended to make things easier and more intuitive. See https://mariadb.com/kb/en/authentication-from-mariadb-104.
Added to release notes.

Izorkin replied:
Added comment in commit.

flokli commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -61,6 +61,19 @@
+      from MariaDB 10.3 to MariaDB 10.4 with Galera Cluster</link> instead.
+      Before doing the upgrade read <link xlink:href="https://mariadb.com/kb/en/upgrading-from-mariadb-103-to-mariadb-104/#incompatible-changes-between-103-and-104">Incompatible
+      Changes Between 10.3 and 10.4</link>.
+      MariaDB 10.4 introduces a number of changes to the authentication process, intended to make things easier and more
+      intuitive. See <link xlink:href="https://mariadb.com/kb/en/authentication-from-mariadb-104/">Authentication from MariaDB 10.4</link>.
```
As a NixOS user upgrading to 20.09, I still don't grok what these "number of changes to the authentication process" now effectively mean for me.

I assume I can't do `sudo -u testuser mysql -u testuser -e 'use testdb; select test_id from db1;' -N` anymore, which is why you updated the test. My question is, was this already the case with mariadb 10.3.22?

flokli replied:
People need to be able to reason about the changes being done, and implications this may have on their configuration.

Izorkin replied:
These changes appeared in 10.4.
Not worked this command:
```
sudo -u testuser mysql -u root
```
Error:
```
ERROR 1698 (28000): Access denied for user 'root'@'localhost'
```
By default, `mysql_install_db` creates two user and used `unix_socket` authentication:
```
CREATE USER root@localhost IDENTIFIED VIA unix_socket OR mysql_native_password USING 'invalid'
CREATE USER mysql@localhost IDENTIFIED VIA unix_socket OR mysql_native_password USING 'invalid'
```

flokli replied:
"It doesn't work anymore, but if I change it to $this, it works again" isn't a sufficient explanation, sorry.

We need to assess the changes and the implications this has on people making use of it, and giving an appropriate heads-up in the release notes at least. If the tests needed updates, it's likely users configuration need changes, too.

Izorkin replied:
When updating, the old authentication method is used. With a new installation it is used `unix_socket` authentication.

Izorkin replied:
@flokli update release notes.

flokli commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -61,6 +61,28 @@
+'';
+</programlisting>
+      When MariaDB data directory is just upgraded (not initialized), the users are not created or modified.
+      After the upgrade you will need to run <literal>mysql_upgrade</literal>.
```
I'd move that line after line 71: ("Before doing the upgrade read <link xlink:href="https://mariadb.com/kb/en/upgrading-from-mariadb-103-to-mariadb-104/#incompatible-changes-between-103-and-104">Incompatible Changes Between 10.3 and 10.4</link>.").

Running `mysql_upgrade` is what all upgrading need to do for sure, so it should come before the authentication parts.

Izorkin replied:
Fixed.

flokli reviewed (changes requested):
A minor nitpick, otherwise should be fine for me now! Thanks for your work!

Izorkin commented:
@flokli @aanderse need add this options - `skip-auth-anonymous-user` and `skip-test-db`?
```
diff --git a/nixos/modules/services/databases/mysql.nix b/nixos/modules/services/databases/mysql.nix
index 44183788d93..582bd1fad82 100644
--- a/nixos/modules/services/databases/mysql.nix
+++ b/nixos/modules/services/databases/mysql.nix
@@ -362,7 +362,7 @@ in

         preStart = if isMariaDB then ''
           if ! test -e ${cfg.dataDir}/mysql; then
-            ${mysql}/bin/mysql_install_db --defaults-file=/etc/my.cnf ${mysqldOptions}
+            ${mysql}/bin/mysql_install_db --defaults-file=/etc/my.cnf ${mysqldOptions} --skip-auth-anonymous-user --skip-test-db
             touch /tmp/mysql_init
           fi
         '' else ''

```

Izorkin commented:
@GrahamcOfBorg test mariadb-galera-mariabackup
@GrahamcOfBorg test mariadb-galera-rsync
@GrahamcOfBorg test mysql
@GrahamcOfBorg test mysql-autobackup
@GrahamcOfBorg test mysql-backup
@GrahamcOfBorg test mysql-replication

Izorkin committed:
mysql57, mysql80: build with protobuf 3.7

They fail to build with protobuf 3.8:

In file included from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol/message.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol_encoder.h:23,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/client_session.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_session.h:25,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_plugin.cc:23:
/build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs_common/protocol_protobuf.h:32:10: fatal error: google/protobuf/wire_format_lite_inl.h: No such file or directory
   32 | #include <google/protobuf/wire_format_lite_inl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [rapid/plugin/x/CMakeFiles/mysqlx.dir/build.make:203: rapid/plugin/x/CMakeFiles/mysqlx.dir/src/xpl_plugin.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....

Izorkin commented:
@GrahamcOfBorg test mariadb-galera-mariabackup
@GrahamcOfBorg test mariadb-galera-rsync
@GrahamcOfBorg test mysql
@GrahamcOfBorg test mysql-autobackup
@GrahamcOfBorg test mysql-backup
@GrahamcOfBorg test mysql-replication

flokli reviewed (commented):
Why is this now built with another protobuf? I'm missing the explanation here.

Izorkin commented:
With protobuf 3.8 mysql 5.7 anf 8.0 error build:

```
In file included from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol/message.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol_encoder.h:23,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/client_session.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_session.h:25,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_plugin.cc:23:
/build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs_common/protocol_protobuf.h:32:10: fatal error: google/protobuf/wire_format_lite_inl.h: No such file or directory
   32 | #include <google/protobuf/wire_format_lite_inl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [rapid/plugin/x/CMakeFiles/mysqlx.dir/build.make:203: rapid/plugin/x/CMakeFiles/mysqlx.dir/src/xpl_plugin.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....

```

flokli commented:
Please name the last commit message like this:

```
mysql57, mysql80: build with protobuf 3.7

They fail to build with protobuf 3.8:

In file included from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol/message.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/protocol_encoder.h:23,
                 from /build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs/client_session.h:24,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_session.h:25,
                 from /build/mysql-5.7.27/rapid/plugin/x/src/xpl_plugin.cc:23:
/build/mysql-5.7.27/rapid/plugin/x/ngs/include/ngs_common/protocol_protobuf.h:32:10: fatal error: google/protobuf/wire_format_lite_inl.h: No such file or directory
   32 | #include <google/protobuf/wire_format_lite_inl.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [rapid/plugin/x/CMakeFiles/mysqlx.dir/build.make:203: rapid/plugin/x/CMakeFiles/mysqlx.dir/src/xpl_plugin.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

Izorkin commented:
Updated and rebased.

flokli commented:
Built all `mysql`/`mariadb` attributes I could find, and ran the `mysql`, `mysql-replication` and `mysql-autobackup` tests.
